### PR TITLE
Don't include proxy requests to KubeAPIErrorBudgetBurn calculation

### PR DESCRIPTION
**Pull request by dohnto**
Address #720

---
There are some [kubernetes API proxy operations](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#-strong-proxy-operations-pod-v1-core-strong-):

**Pods**
 * `POST /api/v1/namespaces/{namespace}/pods/{name}/proxy`
 * `POST /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path}`
 * `DELETE /api/v1/namespaces/{namespace}/pods/{name}/proxy`
 * `DELETE /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path}`
 * `GET /api/v1/namespaces/{namespace}/pods/{name}/proxy`
 * `GET /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path}`
 * `HEAD /api/v1/namespaces/{namespace}/pods/{name}/proxy`
 * `HEAD /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path}`
 * `PUT /api/v1/namespaces/{namespace}/pods/{name}/proxy`
 * `PUT /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path}`

**Services**
 * `POST /api/v1/namespaces/{namespace}/services/{name}/proxy`
 * `POST /api/v1/namespaces/{namespace}/services/{name}/proxy/{path}`
 * `DELETE /api/v1/namespaces/{namespace}/services/{name}/proxy`
 * `DELETE /api/v1/namespaces/{namespace}/services/{name}/proxy/{path}`
 * `GET /api/v1/namespaces/{namespace}/services/{name}/proxy`
 * `GET /api/v1/namespaces/{namespace}/services/{name}/proxy/{path}`
 * `HEAD /api/v1/namespaces/{namespace}/services/{name}/proxy`
 * `HEAD /api/v1/namespaces/{namespace}/services/{name}/proxy/{path}`
 * `PUT /api/v1/namespaces/{namespace}/services/{name}/proxy`
 * `PUT /api/v1/namespaces/{namespace}/services/{name}/proxy/{path}`

**Nodes**
 * `POST /api/v1/nodes/{name}/proxy`
 * `POST /api/v1/nodes/{name}/proxy/{path}`
 * `DELETE /api/v1/nodes/{name}/proxy`
 * `DELETE /api/v1/nodes/{name}/proxy/{path}`
 * `GET /api/v1/nodes/{name}/proxy`
 * `GET /api/v1/nodes/{name}/proxy/{path}`
 * `HEAD /api/v1/nodes/{name}/proxy`
 * `HEAD /api/v1/nodes/{name}/proxy/{path}`
 * `PUT /api/v1/nodes/{name}/proxy`
 * `PUT /api/v1/nodes/{name}/proxy/{path}`

Highlevel description of what proxy means in this context is [here](https://kubernetes.io/docs/concepts/cluster-administration/proxies/):

> The apiserver proxy:
> * is a bastion built into the apiserver
> * connects a user outside of the cluster to cluster IPs which otherwise might not be reachable
> * runs in the apiserver processes
> * client to proxy uses HTTPS (or http if apiserver so configured)
> * proxy to target may use HTTP or HTTPS as chosen by proxy using available information
> * can be used to reach a Node, Pod, or Service
> * does load balancing when used to reach a Service

Because the duration of the HTTP request AND status code is highly dependend on the proxied resource (e. g. getting logs from nodes resource is time heavy, testing Pod's endpoint might result in 5xx status code) I am convinced, that these shoudn't be included in the `KubeAPIErrorBudgetBurn` calculation.

## Open questions

Our environment is affected by the `KubeAPIErrorBudgetBurn` false alert, therefore this PR. But should other rules/alerts be updated as well?

 * [kube-apiserver-histogram.rules](https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/9821d07e94e9a9916575a234fb699ae3331fa939/rules/kube_apiserver.libsonnet#L110) - add the `subresource` label?
 * [apiserver_request:availability*](https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/9821d07e94e9a9916575a234fb699ae3331fa939/rules/kube_apiserver.libsonnet#L249) - the proxy requests "inherit" the status code of the proxied resource. We might be including false 5xx status codes into the calculations. On the other hand if we exclude them, we might be missing some true issues.
 * ...

## Out of scope

Addressing port-forward problem: https://github.com/kubernetes-monitoring/kubernetes-mixin/issues/652

### CODE REVIEW

File: rules/kube_apiserver.libsonnet [hunk 1]
```diff
@@ -4,6 +4,7 @@
     podLabel: 'pod',
     kubeApiserverReadSelector: 'verb=~"LIST|GET"',
     kubeApiserverWriteSelector: 'verb=~"POST|PUT|PATCH|DELETE"',
+    kubeApiserverNonProxySelector: 'subresource!="proxy"',
```
[the remark this answers is not in the document]
Comment (dohnto) [reply rules/kube_apiserver.libsonnet:7]: @aojea I wanted to keep the scope limited to selfishly solve my own problem, but I can extend it if you think it should be covered here.

[the remark this answers is not in the document]
Comment (aojea) [reply rules/kube_apiserver.libsonnet:7]: I think that we should address all the others subresources too but is better to hear from the maintainers on how do they want to proceed

**Comment by arajkumar** [comment]
I'm not a maintainer, but this seem like a reasonable thing to do.

File: rules/kube_apiserver.libsonnet (hunk 1 above)
[the remark this answers is not in the document]
Comment (povilasv) [reply rules/kube_apiserver.libsonnet:7]: Yeah it's better to address the whole problem.

**Review by povilasv — COMMENTED** [review]
Official SLI /SLO states, we should be alerting on *non-streaming*  API calls for reads:

> Latency of non-streaming read-only API calls for every (resource, scope4) pair, measured as 99th percentile over last 5 minutes

SO I think the change should add `kubeApiserverReadNonStreamingSelector` and exclude `proxy|attach|log|exec|portforward`.

Official doc -> https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md

**Comment by dgrisonnet** [comment]
Friendly ping @dohnto

**Comment by dohnto** [comment]
Hi, I have the implementation ready but got stucked with the tests. It is fairly high on my TODO list but not the highest priority as this is just a warning alerts for us.

I more than happy if someone with more resources could step in and fix that.

**Comment by dgrisonnet** [comment]
Thank you for the update @dohnto, I'll open a PR based on your original work and address the comments.

**Comment by dohnto** [comment]
closing as duplicate of https://github.com/kubernetes-monitoring/kubernetes-mixin/pull/740

**Pull request closed without merging**